devito/interfaces.py

```diff
@@ -62,35 +63,42 @@ def __new__(cls, *args, **kwargs):
         return Function.__new__(cls, *args)
 
     def __init__(self):
-        """Initialise from a cached instance by shallow copying __dict__"""
+        """Initialise from a cached instance by shallow copying __dict__."""
         original = _SymbolCache[self.__class__]
         self.__dict__ = original.__dict__.copy()
 
     @classmethod
     def _cached(cls):
-        """Test if current class is already in the symbol cache"""
+        """Test if current class is already in the symbol cache."""
         return cls in _SymbolCache
 
     @classmethod
     def _cache_put(cls, obj):
-        """Store given object instance in symbol cache"""
+        """Store given object instance in symbol cache.
+
+        :param obj: Object to be cached.
+        """
         _SymbolCache[cls] = obj
 
     @classmethod
     def indices(cls, shape):
-        """Abstract class method to determine the default dimension indices."""
+        """Abstract class method to determine the default dimension indices.
+
+        :param shape: Given shape of the data.
```

Comment:
There is a subtle point here that `shape` needs to be the "spatial shape of the data", not the "shape of the data". The subtlety comes from `TimeData` objects actually getting an exta time dimension added automatically on creation, so that if one uses this with the shape of an already instantiated `TimeData` object, we get the wrong indices. We probably want to fix/change that at some point, but this goes beyond the scope of this PR.